tiledb/sm/storage_manager/context.cc

```diff
@@ -50,10 +50,13 @@ Context::Context(const Config& config)
     : last_error_(nullopt)
     , logger_(make_shared<Logger>(
           HERE(), logger_prefix_ + std::to_string(++logger_id_)))
-    , compute_tp_(get_compute_thread_count(config))
-    , io_tp_(get_io_thread_count(config))
     , stats_(make_shared<stats::Stats>(HERE(), "Context"))
```

Comment:
We should put the constructor for the `stats::Stats` object into the `ContextResources` constructor and remove it from here. For the time being it still needs to be `shared_ptr` because that's mandated by `GlobalStats::register_stats`. It's the only thing we need the `shared_ptr` for, so everywhere else we can return a reference to that object.

The constructor argument for `stats_`, here `"Context"`, should be a constructor argument for `ContextResources`, just as you've done with thread pool sizes.

Reply:
There's a bit of a wrinkle there in that each of `Context` and `StorageManager` have different names. I kept the "root" stats object on `Context` so that the `ContextResources` class could keep the "child" instance named the same for backwards compatibility.

Granted, I have no idea what if any contract we have around the JSON stats data but if anyone is using them for monitoring it'd be a bit of a surprise to have them change in a minor version update. Granted, if [2] is up to date and that's our list of stats, we could switch things up to maybe create the root stats object in `ContextResources` with a root name of `Context.StorageManager` that would get renamed whenever 3.0 is released.

[1] https://github.com/TileDB-Inc/TileDB/blob/ed13df7ccf05cae13bec8f3c38cc11c8b7b609ca/tiledb/sm/storage_manager/storage_manager.cc#L95
[2] https://docs.tiledb.com/main/how-to/performance/using-performance-statistics

Reply:
How does this [1] look to you as a solution? Basically, the `Context::stats()` is only ever called once to immediately dump the stats as JSON. Removing that function entirely and replacing the call in the C API with an extra hop through the `ContextResources` instance works just fine. The only minor bit of dirtiness is that I'm labeling the "root" Stats instance as `"Context.StorageManager"` [2] even though its on `ContextResources`. But given the `"Context.StorageManager"` string is technically public API, I think this is the least bad way to handle the situation for an immediate merge with a note that we'll need to figure out a deprecation plan to rename those values.

[1] https://github.com/TileDB-Inc/TileDB/pull/3660/commits/ef49a3f13c057c6fcc3e65e19bcd458cd9d725c7
[2] https://github.com/TileDB-Inc/TileDB/pull/3660/commits/ef49a3f13c057c6fcc3e65e19bcd458cd9d725c7#diff-c8a45c0078699692c264810737e0149c7fd4539f2b7cb3fdc845f4c794ecd3c9R56